mathesar_ui/src/api/tables/records.ts

```diff
@@ -0,0 +1,46 @@
+export type ResultValue = string | number | boolean | null;
+
+export interface Result {
+  [k: string]: ResultValue,
+}
+
+export type GroupingMode = 'distinct' | 'percentile';
+
+export interface Group {
+  /**
+   * The total number of records in the group, even if the group extends beyond
+   * the current page of records.
+   */
+  count: number,
+  first_value: ResultValue,
+  /**
+   * When GroupingMode is 'distinct', then `first_value` and `last_value` will
+   * be identical. Separate first and last values are useful when GroupingMode
+   * is 'percentile'. In that case, the first and last values refer to the
+   * values in the order used for window function in the ranged grouping (which
+   * may not be the same as the order used to sort the result set).
+   */
+  last_value: ResultValue,
+  /**
+   * Each number refers to the index of a record in the response result. This
+   * array will only indices for records returned on the page. If the page is
+   * too small to show all the records, some indices will be missing here.
+   */
+  result_indices: number[],
+}
+
+export interface Grouping {
+  /** Each string is a column name */
+  columns: string[],
+  mode: GroupingMode,
+  /** TODO clarify purpose and behavior from back end team */
+  num_groups: unknown,
```

Comment:
@mathemancer In this file I did my best to capture the typing of the API behavior but I wasn't sure about the `num_groups` field. For example, I'm observing it as `null` when my response has 18 groups. I'm doing distinct grouping though. Maybe it's only relevant for percentile grouping? 

For the sake of the front end functionality, there is no need to make any further changes here because we're not doing anything with that field. But for the sake of maintaining good docs and types, it might be nice to clarify in the code here.

Reply:
`num_groups` is specified in the (admittedly long and rambling) main PR comment.  It's a field that you need to set when using the percentile grouping mode.  For that mode, you need to request a specific number of groups (this is why it's not "count"; it's set by the request, not read from the results).  It should be a positive integer whenever it's set.  Setting it when you request `distinct` grouping has no effect.

Reply:
Thanks @mathemancer -- sorry I missed that description. I've fixed this in 7a8ab4f2922ea4ed4023687c41b991a5a533fa41